### PR TITLE
Introduction of bidirectional calculation

### PR DESCRIPTION
**Pull request by dlohmeier**
Replaces #615 (better cooperation of retoflow personnel on this branch).

### CODE REVIEW

File: src/pandapipes/component_models/heat_consumer_component.py [hunk 1]
```diff
@@ -65,12 +63,15 @@ def create_pit_branch_entries(cls, net, branch_pit):
...
-        hs_pit[:, QEXT] = net[cls.table_name()].qext_w.values
-        return hs_pit
+        hc_pit = super().create_pit_branch_entries(net, branch_pit)
+        hc_pit[:, D] = net[cls.table_name()].diameter_m.values
+        hc_pit[:, AREA] = hc_pit[:, D] ** 2 * np.pi / 4
+        hc_pit[:, MDOTINIT] = net[cls.table_name()].controlled_mdot_kg_per_s.values
+        hc_pit[:, QEXT] = net[cls.table_name()].qext_w.values
+        # causes otherwise problems in case of mode Q
+        hc_pit[np.isnan(hc_pit[:, MDOTINIT]), MDOTINIT] = 0.1
+        hc_pit[hc_pit[:, QEXT] == 0, ACTIVE] = False
```
Comment (dlohmeier) [inline src/pandapipes/component_models/heat_consumer_component.py:73]: Why do we need this? What about mdot and t_return provided?

File: src/pandapipes/component_models/heat_consumer_component.py [hunk 2]
```diff
@@ -119,50 +137,60 @@ def adaption_after_derivatives_hydraulic(cls, net, branch_pit, node_pit, idx_loo
...
-    #     f, t = idx_lookups[cls.table_name()]
-    #     hs_pit = branch_pit[f:t, :]
-    #     from_nodes = hs_pit[:, FROM_NODE_T].astype(np.int32)
-    #
-    #     mask_qext = ~np.isnan(hs_pit[:, QEXT])
-    #     mask_deltat = ~np.isnan(hs_pit[:, DELTAT])
-    #     mask_t_return = ~np.isnan(hs_pit[:, TRETURN])
-    #     mask_mass = ~np.isnan(hs_pit[:, MASS])
-    #     hs_pit[mask_t_return | mask_deltat, JAC_DERIV_DT1] = 0
+        consumer_array = get_component_array(net, cls.table_name())
```
Comment (dlohmeier) [inline src/pandapipes/component_models/heat_consumer_component.py:140]: We should add a test with some consumers oos and different control strategies... I am wondering if this is working properly, as the hc_pit here is the active pit.

Comment (dlohmeier) [reply src/pandapipes/component_models/heat_consumer_component.py:140]: Ah, just saw that the default is get_component_array only refers to active pit

File: src/pandapipes/component_models/heat_consumer_component.py [hunk 3]
```diff
@@ -119,50 +137,60 @@ def adaption_after_derivatives_hydraulic(cls, net, branch_pit, node_pit, idx_loo
...
+        hc_pit[:, JAC_DERIV_DM] = 1
+        hc_pit[:, LOAD_VEC_BRANCHES] = 0
+
+        mask = consumer_array[:, cls.MODE] == cls.QE_TR
+        if np.any(mask):
+            cp = get_branch_cp(net, get_fluid(net), node_pit, hc_pit)
+            from_nodes = hc_pit[mask, FROM_NODE_T].astype(int)
+            t_out = consumer_array[mask, cls.TRETURN]
+            t_mask = hc_pit[mask, TOUTINIT] == node_pit[from_nodes, TINIT]
+            node_pit[from_nodes[t_mask], TINIT] += 10
```
Comment (dlohmeier) [inline src/pandapipes/component_models/heat_consumer_component.py:154]: Why do we need this?

Comment (dlohmeier) [reply src/pandapipes/component_models/heat_consumer_component.py:154]: If this is needed for initialization, how can we make sure that this number will not lead to instability?

File: src/pandapipes/component_models/heat_consumer_component.py [hunk 4]
```diff
@@ -119,50 +137,60 @@ def adaption_after_derivatives_hydraulic(cls, net, branch_pit, node_pit, idx_loo
...
+            q_ext = cp[mask] * consumer_array[mask, cls.MASS] * consumer_array[mask, cls.DELTAT]
+            hc_pit[mask, QEXT] = q_ext
+
+        mask = consumer_array[:, cls.MODE] == cls.MF_TR
+        if np.any(mask):
+            cp = get_branch_cp(net, get_fluid(net), node_pit, hc_pit)
+            from_nodes = hc_pit[mask, FROM_NODE_T].astype(int)
+            t_in = node_pit[from_nodes, TINIT]
+            t_out = hc_pit[mask, TOUTINIT]
+            q_ext = cp[mask] * consumer_array[mask, cls.MASS] * (t_in - t_out)
```
Comment (dlohmeier) [inline src/pandapipes/component_models/heat_consumer_component.py:178]: We would assume that `consumer_array[mask, cls.MASS]` is the same as `hc_pit[mask, MDOTINIT]`, right? Do we need some assertion for this?

File: src/pandapipes/component_models/heat_consumer_component.py [hunk 5]
```diff
@@ -119,50 +137,60 @@ def adaption_after_derivatives_hydraulic(cls, net, branch_pit, node_pit, idx_loo
...
+            hc_pit[mask, QEXT] = q_ext
```
Comment (dlohmeier) [inline src/pandapipes/component_models/heat_consumer_component.py:179]: What is the qext needed for? Could this be done when extracting results?

File: src/pandapipes/component_models/heat_consumer_component.py [hunk 6]
```diff
@@ -224,5 +246,6 @@ def extract_results(cls, net, options, branch_results, mode):
         """
         required_results_hyd, required_results_ht = standard_branch_wo_internals_result_lookup(net)
 
-        extract_branch_results_without_internals(net, branch_results, required_results_hyd,
-                                                 required_results_ht, cls.table_name(), mode)
+        extract_branch_results_without_internals(
+            net, branch_results, required_results_hyd, required_results_ht, cls.table_name(), mode
+        )
```
Comment (dlohmeier) [inline src/pandapipes/component_models/heat_consumer_component.py:251]: Do we need to extract qext here? Should be rather important for heat consumers with return temperature control?

File: src/pandapipes/create.py [hunk 7]
```diff
@@ -528,8 +519,8 @@ def create_pipe_from_parameters(net, from_junction, to_junction, length_km, diam
     return index
 
 
-def create_valve(net, from_junction, to_junction, diameter_m, opened=True, loss_coefficient=0,
-                 name=None, index=None, type='valve', **kwargs):
+def create_valve(net, from_junction, to_junction, diameter_m, opened=True, loss_coefficient=0, name=None, index=None,
```
Comment (dlohmeier) [inline src/pandapipes/create.py:522]: Is it really possible to change the line lengths here? Maybe we should start thinking about formatters or rules in codacy that check these specifics.

**Review by SimonRubenDrauz — APPROVED** [review]

**Pull request merged**